### PR TITLE
alias heapq.nlargest to topk

### PR DESCRIPTION
**Pull request by mrocklin**
Is this worth the alias?  Keep the name `nlargest`?

```Python
In [16]: topk(2, xrange(int(1e9)))
Out[16]: [999999999, 999999998]
```

### CODE REVIEW

**Comment by mrocklin** [comment]
Hrm, I think I'm now leaning towards staying with `nlargest`.  It's not the name I would have chosen but but keeping it seems less invasive.

**Comment by eriknw** [comment]
I would also prefer `topk` if `nlargest` wasn't already used in heapq.

One concern is that the `key` keyword is available in Python 2, but not Python 3.

**Comment by eriknw** [comment]
I'm +0.01 for `topk`, because it's what I'm used to from the literature.  I also think it needs to have a `key` keyword in python 2 and 3, so simply aliasing `heapq.nlargest` won't work.

**Comment by mrocklin** [comment]
Implemented

File: toolz/itertoolz.py [hunk 1]
```diff
@@ -855,3 +855,24 @@ def diff(*seqs, **kwargs):
...
+
+    Use a key function to change sorted order
+
+    >>> topk(2, ['Alice', 'Bob', 'Charlie', 'Dan'], key=len)
+    ('Charlie', 'Alice')
+    """
+    if key is None:
+        return tuple(heapq.nlargest(k, seq))
+    else:
+        return tuple(pluck(1, heapq.nlargest(k, zip(map(key, seq), seq))))
```
Comment (mrocklin) [inline toolz/itertoolz.py]: Whoops.  This doesn't work if seq is consumable.

File: toolz/itertoolz.py [hunk 2]
```diff
@@ -855,3 +855,25 @@ def diff(*seqs, **kwargs):
...
+
+    Use a key function to change sorted order
+
+    >>> topk(2, ['Alice', 'Bob', 'Charlie', 'Dan'], key=len)
+    ('Charlie', 'Alice')
+    """
+    if key is None:
+        return tuple(heapq.nlargest(k, seq))
+    else:
+        keyed_seq = map(lambda x: (key(x), x), seq)
```
Comment (mrocklin) [inline toolz/itertoolz.py]: Fixed.  Better solutions welcome.

Comment (eriknw) [inline toolz/itertoolz.py]: That looks like a good solution to me.

How about adding:
```python
if not callable(key):
    key = getter(key)
```

Comment (mrocklin) [inline toolz/itertoolz.py]: Done.

**Comment by mrocklin** [comment]
Good to go?

File: toolz/itertoolz.py [hunk 3]
```diff
@@ -855,3 +855,25 @@ def diff(*seqs, **kwargs):
...
+    Use a key function to change sorted order
+
+    >>> topk(2, ['Alice', 'Bob', 'Charlie', 'Dan'], key=len)
+    ('Charlie', 'Alice')
+    """
+    if key is None:
+        return tuple(heapq.nlargest(k, seq))
+    if not callable(key):
+        key = getter(key)
+    keyed_seq = map(lambda x: (key(x), x), seq)
```
Comment (eriknw) [inline toolz/itertoolz.py]: When a key function is passed, is it expected that items with equal key values will be in the same order as the original sequence?  This is not currently the case.  We would need to change `(key(x), x)` to something like `(key(x), i, x)` where `i` is incrementing.

**Comment by mrocklin** [comment]
So I was looking at `heapq` to see if it was stable or not and I see that it *does* seem to provide a `key=` keyword argument in both Python 2.7 and 3.4.  

```Python
Type:        function
String form: <function nlargest at 0x7f9a5dc8b840>
File:        /home/mrocklin/Software/anaconda/envs/py34/lib/python3.4/heapq.py
Definition:  heapq.nlargest(n, iterable, key=None)
Source:
def nlargest(n, iterable, key=None):
    """Find the n largest elements in a dataset.

    Equivalent to:  sorted(iterable, key=key, reverse=True)[:n]
    """
```

So perhaps all of this is reinventing a wheel.

It also does appear to be stable

```Python
In [3]: heapq.nlargest(4, (5, 1, 8, 2, 5, 2, 1, 0), key=lambda x: 1)
Out[3]: [5, 1, 8, 2]
```

**Comment by eriknw** [comment]
> So I was looking at heapq to see if it was stable or not and I see that it does seem to provide a key= keyword argument in both Python 2.7 and 3.4.

Oops, my bad.  I must have somehow looked at the wrong function in 3.4 :-/

**Comment by mrocklin** [comment]
OK, relying on nlargest more directly now.  Merging soon if no comment.

**Comment by eriknw** [comment]
Looks good.  Sorry for the misdirection.

TravisCI failure is a docstest comparison of a dict.  It is benign, although I'm surprised this hasn't come up before, because the doctest should probably have `# doctest: +SKIP` pragma.

**Comment by mrocklin** [comment]
I've gone ahead and skipped a number of dict doctests

**Pull request merged**